File: DevEnvironment/Command/src/main/java/command/core/ParameterSet.java
package command.core;

import base.core.CollectionIterator;
import base.core.Conditions;
import base.interfaces.IIterator;
import command.interfaces.IParameter;
import command.interfaces.IParameterSet;
import command.interfaces.IParameterSetMetadata;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The ParameterSet class implements a parameter-set.
 */
public final class ParameterSet implements IParameterSet {
    private final int index;
    private final IParameterSetMetadata metadata;
    private final Map<String, IParameter> parametersMap;
    private final Logger log = LoggerFactory.getLogger(this.getClass());

    /**
     * The ParameterSet constructor.
     */
    public ParameterSet(
        int index,
        IParameterSetMetadata metadata,
        List<IParameter> parameters) {

        Conditions.validateNotNull(
            metadata,
            "The metadata of a parameter-set.");

        Conditions.validateNotNull(
            parameters,
            "The parameters of a parameter-set.");

        this.index = index;
        this.metadata = metadata;
        this.parametersMap = createParameterMap(parameters);
    }

    /**
     * Gets index of a parameter-set.
     */
    @Override
    public int getIndex() {
        return this.index;
    }

    /**
     * Gets metadata of a parameter-set.
     */
    @Override
    public IParameterSetMetadata getMetadata() {
        return this.metadata;
    }

    /**
     * Gets parameters of a parameter-set.
     */
    @Override
    public Map<String, IParameter> getParameters() {
        return this.parametersMap;
    }

    /**
     * Gets a parameter by manifest id.
     */
    @Override
    public IParameter getParameter(String name) {
        return this.parametersMap.get(name);
    }

    /**
     * Checks whether a parameter exists by manifest name.
     */
    @Override
    public boolean hasParameter(String name) {
        return this.parametersMap.containsKey(name);
    }

    /**
     * Gets the iterator.
     */
    @Override
    public IIterator<IParameter> getIterator() {
        return CollectionIterator.make(this.parametersMap.values());
    }

    /**
     * Creates the parameters map.
     */
    private static Map<String, IParameter> createParameterMap(List<IParameter> parameters) {
        Map<String, IParameter> parametersMap = new HashMap<>();

        for (IParameter parameter : parameters) {
            parametersMap.put(parameter.getMetadata().getName(), parameter);
        }

        return parametersMap;
    }
}

File: DevEnvironmentCLion/Command/core/ParameterSet.cpp
#include "PreCompiled.h"
#include "ParameterSet.h"

using namespace command;

/**
 * Creates a new parameter set.
 */
IParameterSetSharedPtr ParameterSet::make(
    int index,
    const std::vector<IParameterSharedPtr>& parameters)
{
    return std::make_shared<ParameterSet>(index, parameters);
}

/**
 * The ParameterSet constructor.
 */
ParameterSet::ParameterSet(
    int index,
    const std::vector<IParameterSharedPtr>& parameters) :
    m_index(index)
{
    createParameterMap(parameters);
}

/**
 * The ParameterSet destructor.
 */
ParameterSet::~ParameterSet()
{
}

/**
 * Gets index of a parameter-set.
 */
int ParameterSet::getIndex() const
{
    return m_index;
}

/**
 * Gets parameters map of a parameter-set.
 */
const ParameterSet::ParameterMap& ParameterSet::getParameters() const
{
    return m_parameters;
}

/**
 * Gets a parameter by manifest name.
 * The parameter can be null.
 */
IParameterSharedPtr ParameterSet::getParameter(const std::string& name) const
{
    ParameterMap::const_iterator parameterIterator = m_parameters.find(name);

    if (parameterIterator == m_parameters.end())
    {
        return nullptr;
    }

    return parameterIterator->second;
}

/**
 * Checks whether a parameter exists by manifest name.
 */
bool ParameterSet::hasParameter(const std::string& name) const
{
    ParameterMap::const_iterator parameterIterator = m_parameters.find(name);
    return parameterIterator != m_parameters.end();
}

/**
 * Creates the parameter map.
 */
void ParameterSet::createParameterMap(const std::vector<IParameterSharedPtr>& parameters)
{
    for (IParameterSharedPtr parameter : parameters)
    {
        m_parameters.insert(std::make_pair(parameter->getName(), parameter));
    }
}
